File: src/types/shot.ts
import type { Character } from './character';

export type CameraAngle = 'closeup' | 'wide' | 'medium' | 'aerial' | 'pan' | 'tilt' | 'tracking' | 'low_angle';

export type ShotStatus = 'pending' | 'images_generated' | 'image_locked' | 'video_generated' | 'completed';

export interface ShotImage {
  id: number;
  shot_id: number;
  image_url: string;
  is_locked: boolean;
  created_at: string;
}

export interface ShotVideo {
  id: number;
  shot_id: number;
  video_url: string;
  source_image_id?: number;
  is_locked: boolean;
  created_at: string;
}

export interface Shot {
  id: number;
  scene_id: number;
  title?: string;
  narration?: string;
  dialogue?: string;
  subtitle?: string;
  action_description?: string;
  camera_angle?: CameraAngle;
  generated_prompt?: string;
  locked_image_id?: number;
  locked_video_id?: number;
  /** 场景参考图 id（来自当前分镜所属场景的 environment.images） */
  ref_environment_image_id?: number | null;
  /** 旧字段：单个角色参考图 id，仅用于向下兼容。新逻辑读 ref_character_view_ids */
  ref_character_view_id?: number | null;
  /** 多角色多参考图：每个被选角色可锁一张 view 作为参考图 */
  ref_character_view_ids?: number[] | null;
  /** 图片提示词模块开关：null 表示全开（兼容旧分镜） */
  prompt_modules_image?: PromptModuleSwitches | null;
  /** 视频提示词模块开关：null 表示全开 */
  prompt_modules_video?: PromptModuleSwitches | null;
  /** 用户自定义图片提示词；非空时覆盖开关拼接结果 */
  custom_prompt_image?: string | null;
  /** 用户自定义视频提示词；非空时覆盖开关拼接结果 */
  custom_prompt_video?: string | null;
  video_url?: string;
  audio_url?: string;
  sort_order: number;
  status: ShotStatus;
  characters: Character[];
  images: ShotImage[];
  videos: ShotVideo[];
  created_at: string;
  updated_at: string;
}

export interface CreateShotRequest {
  title?: string;
  narration?: string;
  dialogue?: string;
  subtitle?: string;
  action_description?: string;
  camera_angle?: CameraAngle;
  character_ids?: number[];
  sort_order?: number;
}

export interface UpdateShotRequest {
  title?: string;
  narration?: string;
  dialogue?: string;
  subtitle?: string;
  action_description?: string;
  camera_angle?: CameraAngle;
  character_ids?: number[];
  /** 场景参考图（来自当前分镜所属场景的 environment.images） */
  ref_environment_image_id?: number | null;
  /** 多角色多参考图：每个被选角色各锁一张 view */
  ref_character_view_ids?: number[] | null;
  /** 提示词模块开关 + 自定义提示词（custom_prompt_* 传 "" 视为清空） */
  prompt_modules_image?: PromptModuleSwitches | null;
  prompt_modules_video?: PromptModuleSwitches | null;
  custom_prompt_image?: string | null;
  custom_prompt_video?: string | null;
}

export interface ShotOrder {
  shot_id: number;
  scene_id: number;
  sort_order: number;
}

/** 提示词模块的 6 个开关键。后端 PROMPT_MODULE_KEYS 与此保持一致。 */
export type PromptModuleKey =
  | 'style'
  | 'environment'
  | 'characters'
  | 'action'
  | 'dialogue'
  | 'camera';

export type PromptModuleSwitches = Record<PromptModuleKey, boolean>;

export interface PromptComponents {
  style: string;
  environment: string;
  characters: string;
  action: string;
  dialogue: string;
  camera: string;
}

export interface PromptPreview {
  prompt: string;
  components: PromptComponents;
  /** True 表示当前用的是自定义提示词，开关已失效 */
  is_custom: boolean;
  /** 当前实际生效的开关状态（None 在响应里被展开为完整 dict） */
  modules: PromptModuleSwitches;
}

export type PromptType = 'image' | 'video';

/** 给前端 ShotEditor 用的 6 个模块的展示元数据 */
export const PROMPT_MODULE_META: Array<{ key: PromptModuleKey; label: string; hint?: string }> = [
  { key: 'style', label: '风格' },
  { key: 'environment', label: '场景' },
  { key: 'characters', label: '角色' },
  { key: 'action', label: '动作' },
  { key: 'dialogue', label: '对白' },
  { key: 'camera', label: '镜头角度' },
];

export const DEFAULT_PROMPT_MODULES: PromptModuleSwitches = {
  style: true,
  environment: true,
  characters: true,
  action: true,
  dialogue: true,
  camera: true,
};

File: backend/app/schemas/shot.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class CharacterBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str


class ShotImageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    shot_id: int
    image_url: str
    is_locked: bool
    created_at: datetime


class ShotVideoOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    shot_id: int
    video_url: str
    source_image_id: Optional[int] = None
    is_locked: bool
    created_at: datetime


class ShotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    scene_id: int
    title: Optional[str] = None
    narration: Optional[str] = None
    dialogue: Optional[str] = None
    subtitle: Optional[str] = None
    action_description: Optional[str] = None
    camera_angle: Optional[str] = None
    generated_prompt: Optional[str] = None
    locked_image_id: Optional[int] = None
    locked_video_id: Optional[int] = None
    ref_environment_image_id: Optional[int] = None
    ref_character_view_id: Optional[int] = None
    ref_character_view_ids: Optional[list[int]] = None
    # 提示词模块开关 + 用户自定义提示词。None 表示"全部启用"（兼容旧分镜）
    prompt_modules_image: Optional[dict[str, bool]] = None
    prompt_modules_video: Optional[dict[str, bool]] = None
    custom_prompt_image: Optional[str] = None
    custom_prompt_video: Optional[str] = None
    video_url: Optional[str] = None
    audio_url: Optional[str] = None
    sort_order: int
    status: str
    characters: list[CharacterBrief] = Field(default_factory=list)
    images: list[ShotImageOut] = Field(default_factory=list)
    videos: list[ShotVideoOut] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class CreateShotRequest(BaseModel):
    title: Optional[str] = None
    narration: Optional[str] = None
    dialogue: Optional[str] = None
    subtitle: Optional[str] = None
    action_description: Optional[str] = None
    camera_angle: Optional[str] = None
    character_ids: list[int] = Field(default_factory=list)
    sort_order: Optional[int] = 0


class UpdateShotRequest(BaseModel):
    title: Optional[str] = None
    narration: Optional[str] = None
    dialogue: Optional[str] = None
    subtitle: Optional[str] = None
    action_description: Optional[str] = None
    camera_angle: Optional[str] = None
    character_ids: Optional[list[int]] = None
    ref_environment_image_id: Optional[int] = None
    ref_character_view_id: Optional[int] = None
    ref_character_view_ids: Optional[list[int]] = None
    # 提示词模块开关 + 自定义提示词。custom_prompt_* 传入 "" 视为清空（回到开关拼接模式）
    prompt_modules_image: Optional[dict[str, bool]] = None
    prompt_modules_video: Optional[dict[str, bool]] = None
    custom_prompt_image: Optional[str] = None
    custom_prompt_video: Optional[str] = None
    sort_order: Optional[int] = None


class ShotOrderItem(BaseModel):
    shot_id: int
    scene_id: int
    sort_order: int


class ReorderShotsRequest(BaseModel):
    shot_orders: list[ShotOrderItem]


class LockImageRequest(BaseModel):
    image_id: int


class LockVideoRequest(BaseModel):
    video_id: int


class PromptComponents(BaseModel):
    style: str = ""
    environment: str = ""
    characters: str = ""
    action: str = ""
    dialogue: str = ""
    camera: str = ""


class PromptPreviewResponse(BaseModel):
    """提示词预览。

    - prompt：最终用于喂给 AI 的字符串
    - components：每个模块的原始片段，供前端展示与开关 UI 使用
    - is_custom：True 表示当前用的是用户自定义提示词，开关已失效
    - modules：当前实际生效的开关状态（None 在响应里会展开为 dict）
    """
    prompt: str
    components: PromptComponents
    is_custom: bool = False
    modules: dict[str, bool] = Field(default_factory=dict)
